### PR TITLE
ONNX Export Interpolate (Resize) for opset version 10

### PR DESCRIPTION
**Pull request by lara-hdr**

### CODE REVIEW

**Comment by lara-hdr** [comment]
waiting on https://github.com/pytorch/pytorch/pull/20533 to be merged

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -667,59 +667,34 @@ def replication_pad(g, input, padding):
...
+        else:
+            scales_constant = [1. if i < 2 else
+                               float(output_size[-(dim - i)]) / float(input.type().sizes()[-(dim - i)])
+                               for i in range(0, dim)]
+            scales = g.op("Constant", value_t=torch.tensor(scales_constant))
+        return g.op("Upsample", input, scales, mode_s=interpolate_mode)
+    return symbolic_fn
+
+
+upsample_nearest1d = _interpolate('upsample_nearest1d', 3, "nearest")
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: fuse upsamples together is great!

maybe we should have some convention on the prefix of the helper functions, such as _interpolate, slice_op, they are not real symbolics, just some helper functions.

**Review by houseroad — COMMENTED** [review]
Rebase to the master?

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (lara-hdr) [reply torch/onnx/symbolic_opset9.py]: I renamed slice to follow the same convention as _interpolate and other ops (_max_pool, _avg_pool).

File: torch/nn/functional.py [hunk 2]
```diff
@@ -2523,9 +2523,9 @@ def _output_size(dim):
 
         # make scale_factor a tensor in tracing so constant doesn't get baked in
         if torch._C._get_tracing_state():
-            return [(torch.floor(input.size(i + 2) * torch.tensor(float(scale_factors[i])))) for i in range(dim)]
+            return [(torch.floor(float(input.size(i + 2)) * torch.tensor(float(scale_factors[i])))) for i in range(dim)]
         else:
-            return [int(math.floor(int(input.size(i + 2)) * scale_factors[i])) for i in range(dim)]
```
Comment (lara-hdr) [inline torch/nn/functional.py]: This cast was making the input of the operator Floor set as "integer", which is not supported (Floor only accepts float and double as input types), and fails in onnxruntime.

File: torch/onnx/symbolic_opset10.py [hunk 3]
```diff
@@ -119,7 +119,34 @@ def symbolic_fn(g, input, kernel_size, stride, padding, ceil_mode, count_include
...
+                               float(output_size[-(dim - i)]) / float(input.type().sizes()[-(dim - i)])
+                               for i in range(0, dim)]
+            scales = g.op("Constant", value_t=torch.tensor(scales_constant))
+        return g.op("Resize", input, scales, mode_s=interpolate_mode)
+    return symbolic_fn
+
+upsample_nearest1d = _interpolate('upsample_nearest1d', 3, "nearest")
+upsample_nearest2d = _interpolate('upsample_nearest2d', 4, "nearest")
+upsample_nearest3d = _interpolate('upsample_nearest3d', 5, "nearest")
+
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset10.py]: Removed support for linear mode in both opsets 9 and 10.
Linear mode behaves differently in onnx and pytorch (with both align_coner enabled and disabled).
The exported model was not matching numbers with onnxruntime.
(the difference is explained here : https://github.com/microsoft/onnxruntime/issues/1179).

Will try to add support to linear mode in a different PR.

File: torch/onnx/symbolic_opset10.py [hunk 4]
```diff
@@ -23,7 +23,7 @@
 # It is very important to blacklist these operators to avoid exporting
 # models with mixed versions of operators.
 # TODO : add support for the blacklisted operators in black_listed_operators
```
Comment (houseroad) [inline torch/onnx/symbolic_opset10.py]: do we want to remove black_listed_operator? Since everything is done :-)

**Comment by lara-hdr** [comment]
@houseroad feedback on this?

**Comment by houseroad** [comment]
I think the failing tests are related. Could you take a look?

File: torch/nn/functional.py (hunk 2 above)
Comment (houseroad) [reply torch/nn/functional.py]: I am a bit confused here, seems the tracer won't go with the else branch, why will it affect onnx's export? do you test the script module? But it's equivalent change, so i am okay with it.

File: torch/nn/functional.py [hunk 5]
```diff
@@ -2523,9 +2523,9 @@ def _output_size(dim):
 
         # make scale_factor a tensor in tracing so constant doesn't get baked in
         if torch._C._get_tracing_state():
-            return [(torch.floor(input.size(i + 2) * torch.tensor(float(scale_factors[i])))) for i in range(dim)]
+            return [(torch.floor((input.size(i + 2) * torch.tensor(float(scale_factors[i]))).float())) for i in range(dim)]
```
Comment (houseroad) [inline torch/nn/functional.py]: seems the old code should generate the float tensor as well, is .float() conversion necessary?

**Review by houseroad — APPROVED** [review]
Looks good. Please address the questions before merge.

File: torch/nn/functional.py (hunk 5 above)
Comment (lara-hdr) [reply torch/nn/functional.py]: Without this, the generated ONNX graph fails in ORT and complains that the input of the "Floor" node is an Int. 
I could submit an example of the ONNX file and the error if required.

For the following line, I did the same for consistency and out of precaution.

Comment (houseroad) [reply torch/nn/functional.py]: Yea, i am wondering how the onnx model looke like? could you attach the exported model?

**Comment by lara-hdr** [comment]
[Resize.zip](https://github.com/pytorch/pytorch/files/3307488/Resize.zip)

@houseroad I attached the model generated without the changes in functional.py, for the code below.
I also added a screenshot of the error I mentioned above.

def Test_Upsample_op():
    class Test_Upsample(nn.Module):
        def forward(self, input):
            return nn.functional.interpolate(input, mode="nearest", scale_factor=2)
    model = Test_Upsample()
    input = torch.tensor([[[[1., 2],[3, 4],]]])
    output = model(input)
    onnx_helper.Save('upsample', 'upsample', model, input, output, 10)

**Pull request closed without merging**